pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,232 @@
+package nodelabeller
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	labelNamespace    = "feature.node.kubernetes.io"
+	labellerNamespace = "node-labeller"
+)
+
+var nodeLabellerVolumePath = "/var/lib/kubevirt-node-labeller"
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	kvmController     *device_manager.DeviceController
+	configMapInformer cache.SharedIndexInformer
+	nodeInformer      cache.SharedIndexInformer
+	clientset         kubecli.KubevirtClient
+	Queue             workqueue.RateLimitingInterface
+	host              string
+	namespace         string
+	logger            *log.FilteredLogger
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, nodeInformer, configMapInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient, host, namespace string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController:     kvmController,
+		configMapInformer: configMapInformer,
+		nodeInformer:      nodeInformer,
+		clientset:         clientset,
+		Queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		host:      host,
+		namespace: namespace,
+		logger:    log.DefaultLogger(),
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	if !n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+		n.logger.Infof("Node-labeller cannot work without KVM device.")
+		return
+	}
+
+	n.configMapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(_ interface{}) {
+			n.Queue.Add(n.host)
+		},
+		DeleteFunc: func(_ interface{}) {
+			n.Queue.Add(n.host)
```

Comment:
I think that we have to handle tombstone objects too here.

Here an example from another place with an explanation:

```
	vmi, ok := obj.(*virtv1.VirtualMachineInstance)

	// When a delete is dropped, the relist will notice a migration in the store not
	// in the list, leading to the insertion of a tombstone object which contains
	// the deleted key/value. Note that this value might be stale. If the migration
	// changed labels the new vmi will not be woken up till the periodic resync.
	if !ok {
		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
		if !ok {
			log.Log.Reason(fmt.Errorf("couldn't get object from tombstone %+v", obj)).Error("Failed to process delete notification")
			return
		}
		vmi, ok = tombstone.Obj.(*virtv1.VirtualMachineInstance)
		if !ok {
			log.Log.Reason(fmt.Errorf("tombstone contained object that is not a migration %#v", obj)).Error("Failed to process delete notification")
			return
		}
	}
```

Reply:
normally yes. but this controller is so dead simple, there's just nothing to look at in the tombstone object. if the node or configmap gets deleted, there's only one key to throw on the queue and it's the same regardless of what happens. 

basically the object passed into the callbacks isn't even useful for this controller. it's kind of confusing at first